clang/lib/Sema/SemaDecl.cpp

```diff
@@ -14277,6 +14325,113 @@ StmtResult Sema::ActOnCXXForRangeIdentifier(Scope *S, SourceLocation IdentLoc,
                                                       : IdentLoc);
 }
 
+static ImplicitConversionKind GetConversionKind(QualType FromType,
```

Comment:
Thanks for the review?
Just to double check that I understand right, are you suggesting to drop the whole approach that reuses C++ narrowing check due to need to keep this `ImplicitConversionKind` construction code, or you're suggesting to somehow find existing code that builds up `StandardConversionSequence` and reuse so there is no additional code that depends on these conversion kinds?

I think I can somehow integrate into `InitializationSequence`, but at the moment I'm having troubles with creating `StandardConversionSequence` object like how it is done there for C++ narrowing check because it has special initialization step which is used only for C++.

Reply:
@AaronBallman , I think I could reuse `IsStandardConversion` to build `StandardConversionSequence` object and call it from `InitializationSequence::Perform`, so `InitializationSequence` will take care of the recursion.

Reply:
That's along the lines of what I was thinking, yes. I was hoping we could use something like `Sema::TryImplicitConversion()` to get us an implicit conversion sequence. From there, we could check to see if it's a standard conversion sequence to get the object which gives access to `StandardConversionSequence::getNarrowingKind()` to see if it's been narrowed.

Reply:
Ok, I've made an attempt to implement that.